src/EditorFeatures/Core/Implementation/Formatting/InferredIndentationDocumentOptionsProviderFactory.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+#nullable enable
+
+using System.ComponentModel.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.Formatting
+{
+    [Export(typeof(IDocumentOptionsProviderFactory))]
+    internal sealed class InferredIndentationDocumentOptionsProviderFactory : IDocumentOptionsProviderFactory
+    {
+        private readonly IIndentationManagerService _indentationManagerService;
+
+        [ImportingConstructor]
+        public InferredIndentationDocumentOptionsProviderFactory(IIndentationManagerService indentationManagerService)
+        {
+            _indentationManagerService = indentationManagerService;
+        }
+
+        public IDocumentOptionsProvider? TryCreate(Workspace workspace)
+        {
+            return new DocumentOptionsProvider(_indentationManagerService);
+        }
+
+        private class DocumentOptionsProvider : IDocumentOptionsProvider
+        {
+            private readonly IIndentationManagerService _indentationManagerService;
+
+            public DocumentOptionsProvider(IIndentationManagerService indentationManagerService)
+            {
+                _indentationManagerService = indentationManagerService;
+            }
+
+            public Task<IDocumentOptions?> GetOptionsForDocumentAsync(Document document, CancellationToken cancellationToken)
+            {
+                return Task.FromResult<IDocumentOptions?>(new DocumentOptions(document.Project.Solution.Workspace, document.Id, _indentationManagerService));
+            }
+
+            private sealed class DocumentOptions : IDocumentOptions
+            {
+                private readonly Workspace _workspace;
+                private readonly DocumentId _documentId;
+                private readonly IIndentationManagerService _indentationManagerService;
+
+                public DocumentOptions(Workspace workspace, DocumentId id, IIndentationManagerService indentationManagerService)
+                {
+                    _workspace = workspace;
+                    _documentId = id;
+                    _indentationManagerService = indentationManagerService;
+                }
+
+                public bool TryGetDocumentOption(OptionKey option, out object? value)
```

Comment:
nit, notnullwhen?

Reply:
Null is a valid value even if returning true.

Reply:
Comment added to the interface.